compiler/src/iree/compiler/Codegen/Common/ExtractAddressComputation.h

```diff
@@ -0,0 +1,84 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+#ifndef IREE_COMPILER_CODEGEN_COMMON_EXTRACTADDRESSCOMPUTATION_H_
+#define IREE_COMPILER_CODEGEN_COMMON_EXTRACTADDRESSCOMPUTATION_H_
+
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/IR/PatternMatch.h"
+
+using namespace mlir;
+
+namespace mlir {
+namespace iree_compiler {
+
+/// Rewrite a load-like op so that all its indices are zeros.
+/// E.g., %ld = memref.load %base[%off0]...[%offN]
+/// =>
+/// %new_base = subview %base[%off0,.., %offN][1,..,1][1,..,1]
+/// %ld = memref.load %new_base[0,..,0] :
+///    memref<1x..x1xTy, strided<[1,..,1], offset: ?>>
+///
+/// `getSrcMemRef` returns the source memref for the given load-like operation.
+///
+/// Using the given rewriter, `rebuildOpFromAddressAndIndices` creates a new
+/// LoadLikeOp that reads from srcMemRef[indices].
+/// The returned operation will be used to replace loadOp.
+template <typename LoadLikeOp, Value (*getSrcMemRef)(LoadLikeOp),
+          LoadLikeOp (*rebuildOpFromAddressAndIndices)(
+              RewriterBase & /*rewriter*/, LoadLikeOp /*loadOp*/,
+              Value /*srcMemRef*/, ArrayRef<Value> /*indices*/)>
+struct LoadLikeOpRewriter : public OpRewritePattern<LoadLikeOp> {
+  using OpRewritePattern<LoadLikeOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(LoadLikeOp loadLikeOp,
+                                PatternRewriter &rewriter) const override {
+    Value srcMemRef = getSrcMemRef(loadLikeOp);
+    auto ldTy = srcMemRef.getType().cast<MemRefType>();
+    unsigned loadRank = ldTy.getRank();
+    // Don't waste compile time if there is nothing to rewrite.
+    if (loadRank == 0) return failure();
+
+    // If our load already has only zeros as indices there is nothing
+    // to do.
+    SmallVector<OpFoldResult> indices =
+        getAsOpFoldResult(loadLikeOp.getIndices());
+    if (std::all_of(indices.begin(), indices.end(),
+                    [](const OpFoldResult &opFold) {
+                      std::optional<int64_t> maybeConstant =
+                          getConstantIntValue(opFold);
+                      return maybeConstant && *maybeConstant == 0;
```

Comment:
nit: you could use `isConstantIntValue` helper:
https://github.com/llvm/llvm-project/blob/fa326c21919faea33b0470b7083dccbd3c1fc3f2/mlir/include/mlir/Dialect/Utils/StaticValueUtils.h#L82